cou/steps/plan.py

```diff
@@ -491,6 +518,59 @@ def _get_archive_data_steps(analysis_result: Analysis, args: CLIargs) -> list[Pr
     return []
 
 
+def _get_set_noout_steps(analysis_result: Analysis, args: CLIargs) -> list[PreUpgradeStep]:
+    """Get set noout steps.
+
+    :param analysis_result: Analysis result
+    :type analysis_result: Analysis
+    :param args: CLI arguments
+    :type args: CLIargs
+    :return: List of pre-upgrade steps.
+    :rtype: list[PreUpgradeStep]
+    """
+    if args.set_noout and args.upgrade_group in {DATA_PLANE, None}:
+        return [
+            PreUpgradeStep(
+                description="Set ceph cluster 'noout' flag before data plane upgrade",
+                coro=ceph.ensure_noout(analysis_result.model, True),
+            )
+        ]
+    if not args.set_noout and args.upgrade_group in {DATA_PLANE, None}:
+        PlanStatus.add_message(
+            "Setting noout during upgrade is optional but recommended. "
+            "You can use `--set-noout` to add this optional step.",
+            "warning",
+        )
+        return [
+            PreUpgradeStep(
+                description="Verify ceph cluster 'noout' is unset",
+                parallel=False,
+                coro=ceph.assert_noout_state(analysis_result.model, False),
+            )
```

Comment:
Yeah, `set noout` is optional during upgrade. But if noout is unset means the cluster is not ready for upgrade because it's not supposed to be `unset noout`. 

> I remember we discussed possibilities of a ceph cluster accidentally having noout set at the beginning of an upgrade, but it feels like in practice this will be confusing. What if the user doesn't want to use --set-noout, but they manually set noout before running cou?

They should use `--set-noout` then, that's the purpose of cou...

> Maybe here this should be more of a warning in the plan: if the cluster has noout set, warn the user ("noout is set, if this was not intentional, please check the cluster"). If the cluster does not have noout set, warn the user ("noout is not set, it's recommended to set noout before running an upgrade and unset it after. COU can manage this automatically given the --set-noout flag.").

First "if" is a pre-sanity check now, rather than warning as explain above.
Second "if" is done

Reply:
> But if noout is unset means the cluster is not ready for upgrade because it's not supposed to be unset noout.

Do you mean "noout is set" here?  This is getting really confusing to talk about with all the set / unset / not unset /  ... :sweat_smile:

Reply:
> First "if" is a pre-sanity check now, 

Can it be a non-blocking warning in the plan, rather than a blocking error during upgrade though?

Reply:
Yes, it can but this was requested feature by Danny

Reply:
I will close the conversation on behalf of @samuelallan72. I have confirmed that the behavior aligns with the specifications discussed in Mattermost.